p2p/src/net/libp2p/connectivity/types.rs

```diff
@@ -0,0 +1,209 @@
+    ConnectionClosed { peer_id: PeerId },
+
+    /// Connection error
+    ConnectionError { addr: Multiaddr, error: P2pError },
+}
+
+macro_rules! peer_info_equal {
+    ($a:ident, $b:ident) => {
+        $a.public_key == $b.public_key
+            && $a.protocol_version == $b.protocol_version
+            && $a.agent_version == $b.agent_version
+            && $a.listen_addrs == $b.listen_addrs
+            && $a.protocols == $b.protocols
+            && $a.observed_addr == $b.observed_addr
+    };
+}
+
+impl PartialEq for BehaviourEvent {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (
+                BehaviourEvent::InboundAccepted {
+                    addr: addr1,
+                    peer_info: peer_info1,
+                },
+                BehaviourEvent::InboundAccepted {
+                    addr: addr2,
+                    peer_info: peer_info2,
+                },
+            ) => addr1 == addr2 && peer_info_equal!(peer_info1, peer_info2),
+            (
+                BehaviourEvent::OutboundAccepted {
+                    addr: addr1,
+                    peer_info: peer_info1,
+                },
+                BehaviourEvent::OutboundAccepted {
+                    addr: addr2,
+                    peer_info: peer_info2,
+                },
+            ) => addr1 == addr2 && peer_info_equal!(peer_info1, peer_info2),
+            (
+                BehaviourEvent::ConnectionClosed { peer_id: peer_id1 },
+                BehaviourEvent::ConnectionClosed { peer_id: peer_id2 },
+            ) => peer_id1 == peer_id2,
+            (
+                BehaviourEvent::ConnectionError {
+                    addr: addr1,
+                    error: error1,
+                },
+                BehaviourEvent::ConnectionError {
+                    addr: addr2,
+                    error: error2,
+                },
+            ) => addr1 == addr2 && error1 == error2,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for BehaviourEvent {}
+
+#[derive(Debug, PartialEq, Eq)]
+/// Events emitted by the `ConnectionManager`
+pub enum ConnectionManagerEvent {
+    Behaviour(BehaviourEvent),
+    Control(ControlEvent),
+}
+
+/// State of a pending connection
+pub enum ConnectionState {
+    /// Outbound connection has been dialed, wait for `ConnectionEstablished` event
+    Dialing,
+
+    /// Connection established for outbound connection
+    OutboundAccepted,
+
+    /// Connection established for inbound connection
+    InboundAccepted,
+
+    /// Connection is active
+    Active,
+
+    /// Connection is in the process of getting closed
+    Closing,
+}
+
+/// Entry for an established and active connections
+pub struct Connection {
+    /// Active address of the remote peer
+    addr: Multiaddr,
+
+    /// Connection state of the remote peer
+    state: ConnectionState,
+
+    /// `IdentifyInfo` of the remote peer
+    peer_info: Option<identify::IdentifyInfo>,
```

Comment:
Why is this optional?

Reply:
Because it's not available yet when the peer is dialing/waiting for the information to be received from remote.

Reply:
This is a concerning part for me in designing communication logic in general as a complex state machine. Can't we instead define two structs and the correct transition over state update over an enum instead of updating the mutable state? You don't have to do this right now, but think about it.

Reply:
That was the original version of the code (both the one that now exists and the first version of this code) but the resulting state machine was IMO even more complicated than this because state transitions made the code to peek into both connection storages and match against their respective states. Having peer information as `None` while the peer is dialing was in my mind a small price to pay to reduce the complexity. But we can revisit this later if this is not acceptable. I'll add a note to my todo list